Library/Homebrew/utils/cp.rb

```diff
@@ -0,0 +1,48 @@
+# typed: true
+# frozen_string_literal: true
+
+require "extend/os/cp"
+require "system_command"
+
+module Utils
+  # Helper functions for interacting with the `cp` command.
+  module Cp
+    class << self
+      sig {
+        params(
+          source:  T.any(String, Pathname, T::Array[T.any(String, Pathname)]),
+          target:  T.any(String, Pathname),
+          sudo:    T::Boolean,
+          verbose: T::Boolean,
+          command: T.class_of(SystemCommand),
+        ).returns(SystemCommand::Result)
+      }
+      def with_attributes(source, target, sudo: false, verbose: false, command: SystemCommand)
+        # On macOS, `cp -p` guarantees to preserve extended attributes (including quarantine
+        # information) in addition to file mode. Other implementations like coreutils does not
+        # necessarily guarantee the same behavior, but that is fine because we don't really need to
+        # preserve extended attributes except when copying Cask artifacts.
+        command.run! "cp", args: ["-p", *extra_flags, *source, target], sudo:, verbose:
```

Comment:
Unless I'm mistaken, on Linux it appears that `FileUtils.cp` already uses `copy_file_range` under the hood so may turn out to actually be quicker on that platform to stick with it over shelling out.

Reply:
I've done a quick experiment on a Linux server and that seems to be the case.

```console
$ dd if=/dev/zero of=payload bs=1024 count=$((256 * 1024))
262144+0 records in
262144+0 records out
268435456 bytes (268 MB, 256 MiB) copied, 0.813365 s, 330 MB/s
$ time cp --reflink=never payload copied
cp --reflink=never payload copied  0.00s user 0.21s system 5% cpu 4.127 total
$ rm copied
$ time ruby -r 'fileutils' -e 'FileUtils.cp("payload", "copied")'
ruby -r 'fileutils' -e 'FileUtils.cp("payload", "copied")'  0.07s user 0.02s system 98% cpu 0.097 total
```

I'm considering to make the `*with_attributes` methods fall back on `FileUtils`, but only in `extend/os/linux`, since `cask/artifact/moved.rb` shouldn't use `FileUtils` and older macOSes wouldn't benefit from the fallback quite much anyway. Does this sound good to you?